// @unipost/sdk/dist/index.mjs
// src/errors.ts
var UniPostError = class extends Error {
  status;
  code;
  constructor(message, status, code) {
    super(message);
    this.name = "UniPostError";
    this.status = status;
    this.code = code;
  }
};
var AuthError = class extends UniPostError {
  constructor(message = "Authentication failed") {
    super(message, 401, "auth_error");
    this.name = "AuthError";
  }
};
var NotFoundError = class extends UniPostError {
  constructor(message = "Resource not found") {
    super(message, 404, "not_found");
    this.name = "NotFoundError";
  }
};
var ValidationError = class extends UniPostError {
  errors;
  constructor(message = "Validation failed", errors = {}) {
    super(message, 422, "validation_error");
    this.name = "ValidationError";
    this.errors = errors;
  }
};
var RateLimitError = class extends UniPostError {
  retryAfter;
  constructor(retryAfter, message = "Rate limit exceeded") {
    super(message, 429, "rate_limit");
    this.name = "RateLimitError";
    this.retryAfter = retryAfter;
  }
};
var PlatformError = class extends UniPostError {
  platform;
  constructor(message, platform) {
    super(message, 502, "platform_error");
    this.name = "PlatformError";
    this.platform = platform;
  }
};
var QuotaError = class extends UniPostError {
  constructor(message = "Monthly quota exceeded") {
    super(message, 403, "quota_exceeded");
    this.name = "QuotaError";
  }
};
function parseApiError(status, body) {
  const msg = body.error?.message || "Unknown API error";
  const code = body.error?.code || "unknown";
  switch (status) {
    case 401:
      return new AuthError(msg);
    case 404:
      return new NotFoundError(msg);
    case 422:
      return new ValidationError(msg, body.error?.errors);
    case 429:
      return new RateLimitError(0, msg);
    case 403:
      if (code === "quota_exceeded") return new QuotaError(msg);
      return new UniPostError(msg, status, code);
    case 502:
      if (body.error?.platform) return new PlatformError(msg, body.error.platform);
      return new UniPostError(msg, status, code);
    default:
      return new UniPostError(msg, status, code);
  }
}

// src/http.ts
var MAX_RETRIES = 2;
var HttpClient = class {
  apiKey;
  baseUrl;
  timeout;
  constructor(options) {
    this.apiKey = options.apiKey;
    this.baseUrl = options.baseUrl;
    this.timeout = options.timeout;
  }
  async request(method, path, options) {
    const url = new URL(path, this.baseUrl);
    if (options?.query) {
      for (const [key, value] of Object.entries(options.query)) {
        if (value !== void 0) {
          url.searchParams.set(key, String(value));
        }
      }
    }
    const headers = {
      Authorization: `Bearer ${this.apiKey}`,
      "Content-Type": "application/json",
      "User-Agent": "@unipost/sdk/0.1.0",
      ...options?.headers
    };
    const init = {
      method,
      headers,
      signal: AbortSignal.timeout(this.timeout)
    };
    if (options?.body !== void 0) {
      init.body = JSON.stringify(options.body);
    }
    let lastError = null;
    for (let attempt = 0; attempt <= MAX_RETRIES; attempt++) {
      try {
        const response = await fetch(url.toString(), init);
        if (response.ok) {
          if (response.status === 204) return void 0;
          return await response.json();
        }
        if (response.status === 429) {
          const retryAfter = parseInt(response.headers.get("Retry-After") || "1", 10);
          if (attempt < MAX_RETRIES) {
            await sleep(retryAfter * 1e3);
            continue;
          }
          throw new RateLimitError(retryAfter);
        }
        const body = await response.json().catch(() => ({}));
        throw parseApiError(response.status, body);
      } catch (err) {
        if (err instanceof RateLimitError && attempt < MAX_RETRIES) {
          await sleep(err.retryAfter * 1e3);
          lastError = err;
          continue;
        }
        throw err;
      }
    }
    throw lastError || new Error("Request failed after retries");
  }
  get(path, query) {
    return this.request("GET", path, { query });
  }
  post(path, body, headers) {
    return this.request("POST", path, { body, headers });
  }
  put(path, body) {
    return this.request("PUT", path, { body });
  }
  delete(path) {
    return this.request("DELETE", path);
  }
};
function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// src/resources/accounts.ts
var Accounts = class {
  constructor(http) {
    this.http = http;
  }
  http;
  /** List all connected social accounts. */
  async list(params) {
    const query = {};
    if (params?.platform) query.platform = params.platform;
    if (params?.externalUserId) query.external_user_id = params.externalUserId;
    if (params?.status) query.status = params.status;
    return this.http.get("/v1/social-accounts", query);
  }
  /** Get a single account by ID. */
  async get(accountId) {
    const res = await this.http.get(`/v1/social-accounts/${accountId}`);
    return res.data;
  }
  /** Check account health status. */
  async health(accountId) {
    const res = await this.http.get(`/v1/social-accounts/${accountId}/health`);
    return res.data;
  }
};

// src/resources/posts.ts
function toSnakeCase(params) {
  const body = {};
  if (params.caption !== void 0) body.caption = params.caption;
  if (params.accountIds) body.account_ids = params.accountIds;
  if (params.mediaUrls) body.media_urls = params.mediaUrls;
  if (params.scheduledAt) body.scheduled_at = params.scheduledAt;
  if (params.status) body.status = params.status;
  if (params.platformPosts) {
    body.platform_posts = params.platformPosts.map((pp) => {
      const entry = { account_id: pp.accountId };
      if (pp.caption !== void 0) entry.caption = pp.caption;
      if (pp.threadPosition !== void 0) entry.thread_position = pp.threadPosition;
      if (pp.firstComment !== void 0) entry.first_comment = pp.firstComment;
      if (pp.mediaIds) entry.media_ids = pp.mediaIds;
      return entry;
    });
  }
  return body;
}
var Posts = class {
  constructor(http) {
    this.http = http;
  }
  http;
  /** Create a new post. */
  async create(params) {
    const body = toSnakeCase(params);
    const headers = {};
    if (params.idempotencyKey) {
      headers["Idempotency-Key"] = params.idempotencyKey;
    }
    const res = await this.http.post("/v1/social-posts", body, headers);
    return res.data;
  }
  /** List posts with optional filters and cursor pagination. */
  async list(params) {
    const query = {};
    if (params?.status) query.status = params.status;
    if (params?.platform) query.platform = params.platform;
    if (params?.from) query.from = params.from;
    if (params?.to) query.to = params.to;
    if (params?.limit) query.limit = params.limit;
    if (params?.cursor) query.cursor = params.cursor;
    return this.http.get("/v1/social-posts", query);
  }
  /** Iterate all posts using async generator (auto-pagination). */
  async *listAll(params) {
    let cursor;
    do {
      const page = await this.list({ ...params, cursor });
      for (const post of page.data) {
        yield post;
      }
      cursor = page.nextCursor;
    } while (cursor);
  }
  /** Get a single post by ID. */
  async get(postId) {
    const res = await this.http.get(`/v1/social-posts/${postId}`);
    return res.data;
  }
  /** Get analytics for a post. */
  async analytics(postId) {
    const res = await this.http.get(`/v1/social-posts/${postId}/analytics`);
    return res.data;
  }
  /** Publish a draft post. */
  async publish(postId) {
    const res = await this.http.post(`/v1/social-posts/${postId}/publish`);
    return res.data;
  }
  /** Cancel a scheduled post. */
  async cancel(postId) {
    const res = await this.http.post(`/v1/social-posts/${postId}/cancel`);
    return res.data;
  }
  /** Bulk create posts (up to 50). */
  async bulkCreate(posts) {
    const body = posts.map(toSnakeCase);
    const res = await this.http.post("/v1/social-posts/bulk", body);
    return res.data;
  }
};

// src/resources/media.ts
var Media = class {
  constructor(http) {
    this.http = http;
  }
  http;
  /** Request a presigned upload URL. */
  async upload(params) {
    const res = await this.http.post("/v1/media/upload", {
      filename: params.filename,
      content_type: params.contentType,
      size_bytes: params.sizeBytes
    });
    return res.data;
  }
  /**
   * Convenience: upload a local file (Node.js only).
   * Reads the file, requests a presigned URL, uploads, and returns the mediaId.
   */
  async uploadFile(filePath) {
    const { readFileSync, statSync } = await import("fs");
    const { basename } = await import("path");
    const stats = statSync(filePath);
    const filename = basename(filePath);
    const ext = filename.split(".").pop()?.toLowerCase() || "";
    const contentType = MIME_TYPES[ext] || "application/octet-stream";
    const { mediaId, uploadUrl } = await this.upload({
      filename,
      contentType,
      sizeBytes: stats.size
    });
    const fileBuffer = readFileSync(filePath);
    await fetch(uploadUrl, {
      method: "PUT",
      body: fileBuffer,
      headers: { "Content-Type": contentType }
    });
    return mediaId;
  }
};
var MIME_TYPES = {
  jpg: "image/jpeg",
  jpeg: "image/jpeg",
  png: "image/png",
  gif: "image/gif",
  webp: "image/webp",
  mp4: "video/mp4",
  mov: "video/quicktime"
};

// src/resources/analytics.ts
var Analytics = class {
  constructor(http) {
    this.http = http;
  }
  http;
  /** Get aggregated analytics rollup. */
  async rollup(params) {
    const query = {
      from: params.from,
      to: params.to,
      granularity: params.granularity,
      group_by: params.groupBy
    };
    const res = await this.http.get("/v1/analytics/rollup", query);
    return res.data;
  }
};

// src/resources/connect.ts
var Connect = class {
  constructor(http) {
    this.http = http;
  }
  http;
  /** Create a Connect session for end-user OAuth. */
  async createSession(params) {
    const res = await this.http.post("/v1/connect/sessions", {
      platform: params.platform,
      external_user_id: params.externalUserId,
      external_user_email: params.externalUserEmail,
      return_url: params.returnUrl
    });
    return res.data;
  }
  /** Get the status of a Connect session. */
  async getSession(sessionId) {
    const res = await this.http.get(`/v1/connect/sessions/${sessionId}`);
    return res.data;
  }
};

// src/resources/users.ts
var Users = class {
  constructor(http) {
    this.http = http;
  }
  http;
  /** List all managed users (Connect). */
  async list() {
    return this.http.get("/v1/users");
  }
  /** Get a single managed user by external_user_id. */
  async get(externalUserId) {
    const res = await this.http.get(`/v1/users/${externalUserId}`);
    return res.data;
  }
};

// src/client.ts
var DEFAULT_BASE_URL = "https://api.unipost.dev";
var DEFAULT_TIMEOUT = 3e4;
var UniPost = class {
  accounts;
  posts;
  media;
  analytics;
  connect;
  users;
  constructor(options = {}) {
    const apiKey = options.apiKey ?? getEnvVar("UNIPOST_API_KEY");
    if (!apiKey) {
      throw new Error(
        "UniPost API key is required. Pass it as `new UniPost({ apiKey })` or set the UNIPOST_API_KEY environment variable."
      );
    }
    const http = new HttpClient({
      apiKey,
      baseUrl: options.baseUrl ?? DEFAULT_BASE_URL,
      timeout: options.timeout ?? DEFAULT_TIMEOUT
    });
    this.accounts = new Accounts(http);
    this.posts = new Posts(http);
    this.media = new Media(http);
    this.analytics = new Analytics(http);
    this.connect = new Connect(http);
    this.users = new Users(http);
  }
};
function getEnvVar(name) {
  if (typeof process !== "undefined" && process.env) {
    return process.env[name];
  }
  if (typeof globalThis !== "undefined" && "Deno" in globalThis) {
    try {
      return globalThis.Deno.env.get(name);
    } catch {
      return void 0;
    }
  }
  return void 0;
}

// src/webhook.ts
async function verifyWebhookSignature(options) {
  const { payload, signature, secret } = options;
  const encoder = new TextEncoder();
  const key = await crypto.subtle.importKey(
    "raw",
    encoder.encode(secret),
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"]
  );
  const payloadBytes = typeof payload === "string" ? encoder.encode(payload) : new Uint8Array(payload);
  const signatureBytes = await crypto.subtle.sign("HMAC", key, payloadBytes);
  const computed = bufferToHex(signatureBytes);
  return timingSafeEqual(computed, signature);
}
function bufferToHex(buffer) {
  return Array.from(new Uint8Array(buffer)).map((b) => b.toString(16).padStart(2, "0")).join("");
}
function timingSafeEqual(a, b) {
  if (a.length !== b.length) return false;
  let result = 0;
  for (let i = 0; i < a.length; i++) {
    result |= a.charCodeAt(i) ^ b.charCodeAt(i);
  }
  return result === 0;
}
export {
  AuthError,
  NotFoundError,
  PlatformError,
  QuotaError,
  RateLimitError,
  UniPost,
  UniPostError,
  ValidationError,
  verifyWebhookSignature
};
